ADMIN: Create separate issue templates for different usecases

ShaharNaveh opened the pull request:
revival of #31551 originally opened by @jschendel 

----

> I recall this being mentioned on one of the core dev calls a few months back, and took inspiration from [rapidsai/cudf](https://github.com/rapidsai/cudf) for the various templates.
> 
> Added templates for the following use cases:
> 
>     * Bug Report
>       
>       * This is largely the same as the current issue template
> 
>     * Documentation Enhancement
> 
>     * Documentation Error
> 
>     * Feature Request
> 
>     * Submit Question
>       
>       * This attempts to direct users to StackOverflow for usage questions
> 
> 
> We could also add a Blank Template for opening an issue without any template provided, but opted not to do that for now, as I don't want to encourage people to bypass these templates. Could certainly add one if there's a consensus that we want this.
> 
> I've created a [local repo](https://github.com/jschendel/pandas-templates/issues) where you can see these templates in action since I couldn't figure out another way to actually display these. Click the "New Issue" button to see what these changes would look like.
> 
>

jschendel committed:
ADMIN: Create separate issue templates for different usecases

Commit:
Merge two documentation templates

mroeschke commented on .github/ISSUE_TEMPLATE/bug_report.md:
````diff
@@ -1,29 +1,34 @@
 
 ```
+
+**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.
````
IMO, this should go before the `Code Sample`

mroeschke commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,29 +1,34 @@
-[this should explain **why** the current behaviour is a problem and why the expected output is a better solution.]
+[this should explain **why** the current behaviour is a problem and why the expected output is a better solution]
 
 **Note**: We receive a lot of issues on our GitHub tracker, so it is very possible that your issue has been posted before. Please check first before submitting so that we do not have to handle and close duplicates!
```
As a suggestion. I think we should change these two ` **Note** ` sections into checkboxes like

```
- [ ] I have checked that issue has not already been reported
- [ ] I have confirmed this bug exists on the latest version of pandas
- [ ] I have confirmed this bug exists on the master branch of pandas
```

Inspired how Numba does their bug reports: https://github.com/numba/numba/issues/4735

mroeschke commented on .github/ISSUE_TEMPLATE/wrong_or_missing_documentation.md:
```diff
@@ -0,0 +1,20 @@
+
+#### Location of the documentation
+
+[this should provide a description of what documentation you believe needs to be fixed/improved]
```
I think we should also encourage a link to the documentation that they want to see improved.

TomAugspurger commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -0,0 +1,28 @@
+
+#### Describe the solution you'd like
+
+[this should provide a description of the feature request, e.g. "`DataFrame.foo` should get a new parameter `bar` that [...]"]
```
I've found asking people to "write a docstring" to be helpful in clarifying what people are asking for.

TomAugspurger commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -0,0 +1,28 @@
+#### Describe the solution you'd like
+
+[this should provide a description of the feature request, e.g. "`DataFrame.foo` should get a new parameter `bar` that [...]"]
+
```
Can you add a header for "API breaking implications" and a prompt asking about how the proposal affects the API.

TomAugspurger reviewed (commented):
Thanks for reviving this.

Commit:
Merge remote-tracking branch 'upstream/master' into issue-template

Commit:
Moved guide for minimial example before the Code sample section

XREF:
https://github.com/pandas-dev/pandas/pull/33187#discussion_r401164088

Commit:
Using checkboxes in the bug report

XREF:
https://github.com/pandas-dev/pandas/pull/33187#discussion_r401165273

Commit:
Newline nits

Commit:
Encouraging users to supply URL of problematic documentation

Commit:
Renamed "wrong_or_missing_documentation" to "documentation_improvement"

Commit:
Encouraging users to write a docstring for desired feature

XREF:
https://github.com/pandas-dev/pandas/pull/33187#discussion_r401226797

Commit:
Added "API breaking implications" header

XREF:
https://github.com/pandas-dev/pandas/pull/33187#discussion_r401227352

MomIsBestFriend commented on .github/ISSUE_TEMPLATE/submit_question.md:
```diff
@@ -0,0 +1,20 @@
+
+#### Question about pandas
+
+**Note**: It's highly recommended that you use [StackOverflow](https://www.stackoverflow.com) for any usage related questions, e.g. "How do I do [...] with pandas?" You are much more likely to receive a quick response to your question on StackOverflow than the GitHub issue tracker. You may also want to search the [pandas tag](https://stackoverflow.com/questions/tagged/pandas) on StackOverflow to  ...
```
@mroeschke I *think* this is very similar to a proposal you made [here](https://github.com/pandas-dev/pandas/pull/33187#discussion_r401165273), would you like me to change that to a checkbox, as well?

mroeschke replied:
Sure great point. I think the 2 checkboxes could be something like

```
- [ ] I have searched the [pandas tag] on StackOverflow for similar questions
- [ ] I have asked my usage related question on [StackOverflow]
```

MomIsBestFriend commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -0,0 +1,33 @@
+
+#### API breaking implications
+
+[this should provide a description of how this feature will affect the API]
```
@TomAugspurger I have added the header in 7fb0a81, although I'm not sure that the headers description is good/informative.

jorisvandenbossche commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,39 @@
+
+- [ ] I have confirmed this bug exists on the latest version of pandas.
+
+- [ ] I have confirmed this bug exists on the master branch of pandas.
```
I don't think we should necessarily expect that users have master built to report a bug

TomAugspurger replied:
Agreed. Perhaps something like

```suggestion
- [ ] (optinal) I have confirmed this bug exists on the master branch of pandas.
```

which makes the others sound more required too.

jorisvandenbossche commented on .github/ISSUE_TEMPLATE/documentation_improvement.md:
```diff
@@ -0,0 +1,23 @@
+---
+
+name: Documentation Improvment
```
```suggestion
name: Documentation Improvement
```

ShaharNaveh committed:
Update .github/ISSUE_TEMPLATE/bug_report.md

Co-Authored-By: Tom Augspurger <TomAugspurger@users.noreply.github.com>

ShaharNaveh committed:
Update .github/ISSUE_TEMPLATE/documentation_improvement.md

Co-Authored-By: Joris Van den Bossche <jorisvandenbossche@gmail.com>

mroeschke commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,39 @@
+
+- [ ] I have confirmed this bug exists on the latest version of pandas.
+
+- [ ] (optinal) I have confirmed this bug exists on the master branch of pandas.
```
`optinal` -> `optional`

mroeschke commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,39 @@
+
+---
+
+- [ ] I have checked that issue has not already been reported.
```
`that issue` -> `that this issue`

mroeschke commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,39 @@
+name: Bug Report
+about: Create a bug report to help us improve pandas
+title: "BUG:"
+labels: ""
```
I'd say we should use this `label:` feature to automatically tag the issue with the appropriate label. For this template `Bug` and similarly for the others.

https://github.community/t5/How-to-use-Git-and-GitHub/Apply-labels-automatically-at-issue-creation/td-p/10719#

MomIsBestFriend replied:
Just making sure before I make the changes, the labels for each file should look like this:

---

bug_report: Bug

---

documentation_improvement: Docs

---

feature_request: Enhancement?

---

submit_question: "Usage Question"

jschendel replied:
I was originally on the fence about automatically adding labels because it could make it harder to distinguish what has been triaged vs. what hasn't, since I currently look for the presence of labels to determine if something has been triaged, and automatically adding labels would make it less clear whether a triage has occurred or not.

A potential workaround here is to also automatically add a new "Needs Triage" label and then remove it as part of the triage process.  Or maybe this is only an issue on my part, in which case I can work around it.

MomIsBestFriend replied:
We don't have a label for feature requests

jschendel replied:
> We don't have a label for feature requests

We typically use the "Enhancement" label for this, so I think what you've proposed above is good.

MomIsBestFriend replied:
> A potential workaround here is to also automatically add a new "Needs Triage" label and then remove it as part of the triage process.

@jschendel I like that idea, have you seen repositories that do this kind of thing?

jschendel replied:
Yeah, I've seen a couple variants of this: [rapidsai/cudf](https://github.com/rapidsai/cudf/blob/fa525bf1baf0619901de0f356d64baa059448326/.github/ISSUE_TEMPLATE/bug_report.md) automatically adds a generic "? - Needs Triage" label, [scikit-learn/scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/a5dfba94ff34b27848a1094c01c6f6a3dbdf1dae/.github/ISSUE_TEMPLATE/bug_report.md) adds a "Bug: triage" just to bug reports.

mroeschke replied:
Numba has a `need triage` label: https://github.com/numba/numba/issues?q=is%3Aissue+is%3Aopen+label%3Aneedtriage

I'd be in favor in autotagging with `<Bug/Doc/Enhancement/...>  + Need Traige`, and then `Need Triage` can be removed once someone has commented (and can be replace with a more specific label like `IO CSV` for example)

I've also been making an effort to tag older issue with one of Bug/Doc/Enhancement/... which will be useful in the future for developing metrics about total number of issues per category.

MomIsBestFriend replied:
@mroeschke ATM ```Need Triage``` label does not exists.
Should I wait for it to be created? or just edit this pull request to auto-assign the label ```Need Triage``` and the label will be created later.

mroeschke replied:
I went ahead and just created a `Needs Triage` label so it should be available now.

MomIsBestFriend replied:
Thank you @mroeschke

Commit:
Replaced notes with checkboxes

XREF:
https://github.com/pandas-dev/pandas/pull/33187#discussion_r401732329

Commit:
Fixed typos/grammer mistakes

XREF:
https://github.com/pandas-dev/pandas/pull/33187#discussion_r401733136
https://github.com/pandas-dev/pandas/pull/33187#discussion_r401733508

Commit:
Merge remote-tracking branch 'upstream/master' into issue-template

Commit:
Add labels to each issue template

Commit:
Added "Needs Triage" label to all templates

Commit:
Merge remote-tracking branch 'upstream/master' into issue-template

jschendel commented on .github/ISSUE_TEMPLATE/submit_question.md:
```diff
@@ -0,0 +1,24 @@
+
+- [ ] I have searched the [pandas tag] on StackOverflow for similar questions.
+
+- [ ] I have asked my usage related question on [StackOverflow].
```
```suggestion
- [ ] I have asked my usage related question on [StackOverflow](https://stackoverflow.com).
```

ShaharNaveh committed:
Update .github/ISSUE_TEMPLATE/submit_question.md

Co-Authored-By: Jeremy Schendel <jschendel@users.noreply.github.com>

ShaharNaveh committed:
Update .github/ISSUE_TEMPLATE/submit_question.md

Co-Authored-By: Jeremy Schendel <jschendel@users.noreply.github.com>

Commit:
Merge remote-tracking branch 'upstream/master' into issue-template

TomAugspurger reviewed (approved):
Looks great!

jreback commented:
thanks @MomIsBestFriend

jreback commented:
it even pre-sets some labels, nice!
